br/pkg/lightning/backend/local/local.go

```diff
@@ -924,30 +946,32 @@ func splitRangeBySizeProps(fullRange Range, sizeProps *sizeProperties, sizeLimit
 	curSize := uint64(0)
 	curKeys := uint64(0)
 	curKey := fullRange.start
+
 	sizeProps.iter(func(p *rangeProperty) bool {
-		if bytes.Equal(p.Key, engineMetaKey) {
+		if bytes.Compare(p.Key, fullRange.start) <= 0 {
 			return true
 		}
+		if bytes.Compare(p.Key, fullRange.end) > 0 {
+			return false
+		}
 		curSize += p.Size
 		curKeys += p.Keys
-		if int64(curSize) >= sizeLimit || int64(curKeys) >= keysLimit {
-			// in case the sizeLimit or keysLimit is too small
-			endKey := p.Key
-			if bytes.Equal(curKey, endKey) {
-				endKey = nextKey(endKey)
-			}
-			ranges = append(ranges, Range{start: curKey, end: endKey})
-			curKey = endKey
+		if bytes.Compare(p.Key, curKey) > 0 && (int64(curSize) >= sizeLimit || int64(curKeys) >= keysLimit) {
+			ranges = append(ranges, Range{start: curKey, end: p.Key})
+			curKey = p.Key
 			curSize = 0
 			curKeys = 0
 		}
 		return true
 	})
 
-	if curKeys > 0 {
-		ranges = append(ranges, Range{start: curKey, end: fullRange.end})
-	} else {
-		ranges[len(ranges)-1].end = fullRange.end
+	if bytes.Compare(curKey, fullRange.end) < 0 {
```

Comment:
If the loop ended with `curKey > fullRange.end` we should still set the last range's endKey as fullRange.end otherwise we may lose some keys

Reply:
It's impossible that `curKey > fullRange.end`,  because `curKey` is always set after ` bytes.Compare(p.Key, fullRange.end) <= 0`